METRICS.md

```diff
@@ -0,0 +1,282 @@
+FROM four_keys.deployments d, d.changes
+LEFT JOIN four_keys.changes c ON changes = c.change_id;
+```
+
+From this base, we want to extract the daily median lead time to change. 
+
+```sql
+SElECT
+day,
+PERCENTILE_CONT(
+  # Ignore automated changes
+  IF(time_to_change_minutes > 0,time_to_change_minutes, NULL), 
+  0.5) # Median
+  OVER (partition by day) median_time_to_change
+FROM
+(SELECT
+ d.deploy_id,
+ TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+ # Time to Change
+ TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+ FROM four_keys.deployments d, d.changes
+ LEFT JOIN four_keys.changes c ON changes = c.change_id
+ )
+GROUP BY day, time_to_change_minutes;
+```
+
+It is up for debate whether or not we want to ignore the automated changes.  The rationale here is that when we merge a Pull Request it creates a Push event in the main branch.  This Push event is not its own distinct change, but rather a link in the workflow.  If we trigger a deployment off of this push event, this artificially skews the metrics and does not give us a clear picture of developer velocity. 
+
+### Calculating the bucket ###
+![Image of chart from the Four Keys dashboard, showing the median Lead Time to Change.](images/lead_time.png)
+
+To get the buckets, rather than aggregating daily, we look at the last 3 months and bucket the results according to the DORA research. 
+
+```sql
+SELECT 
+CASE WHEN median_time_to_change < 24 * 60 then "One day"
+     WHEN median_time_to_change < 168 * 60 then "One week"
+     WHEN median_time_to_change < 730 * 60 then "One month"
+     WHEN median_time_to_change < 730 * 6 * 60 then "Six months"
+     ELSE "One year"
+     END as lead_time_to_change,
+FROM
+  (SElECT
+    PERCENTILE_CONT(
+    # Ignore automated changes
+    IF(time_to_change_minutes > 0,time_to_change_minutes, NULL), 
+    0.5) # Median
+    OVER () median_time_to_change
+      FROM
+      (SELECT
+       d.deploy_id,
+       TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+       # Time to Change
+       TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+       FROM four_keys.deployments d, d.changes
+       LEFT JOIN four_keys.changes c ON changes = c.change_id
+       # Limit to 3 months
+       WHERE d.time_created > TIMESTAMP(DATE_SUB(CURRENT_DATE(), INTERVAL 3 MONTH))
+       )
+      GROUP BY day, time_to_change_minutes
+      )
+LIMIT 1;
+```
+
+## Time to Restore Services ##
+
+**Definition**: For a failure, the median amount of time between the deployment which caused the failure and the restoration.  The restoration is measured by closing an associated bug / incident report. 
+
+### Daily Median Time to Restore Services ###
+![Image of chart from the Four Keys dashboard, showing the daily MTTR.](images/daily_mttr.png)
+
+To calculate the daily median time to restore services, all the information we need is in the incidents table.
```

Comment:
Slightly off topic, but: do we need to aggregate by day _prior_ to feeding these data to Data Studio? What would a graph look like with all the data points? Might it be (modestly) more detailed?

Reply:
Interesting.  Originally, I didn't want to do any pre-aggregating and wanted it all to be in the dashboard so you could drill down.  I couldn't find a good way to get Datastudio to do that at the time though, so went with this daily aggregation approach on all the metrics.  

TBH, it would be nice to also be able to see how the buckets changed over time, since this data is available in the tables.  Would it be worth creating additional pages in the dashboard?

Reply:
Ah, okay, thanks for the detail. IMHO, enriching the dashboard may well be useful, but I think we should allow some time to gather requirements/evidence first.